### PR TITLE
Added overlap accessors to spatial_pooler.py

### LINKED ISSUE #3148: Add overlap accessors to spatial_pooler.py
This issue is to add the following getter's to the Python SpatialPooler:

getOverlapScores
getBoostedOverlapScores

This would be implemented by adding self._overlaps and self._boostedOverlap member variables.  These should be initialized in the constructor to all zeros.

This will be a backward incompatible change, in that previously serialized spatial poolers will not load once this change has been made. As such, we should increment the VERSION flag in spatial_pooler.py and then in __setstate__() we should add the members in for older versions of the SP.

Tests should be added to ensure we are getting the correct results for overlap and boostedOverlap.

A corresponding change needs to be made to the C++ implementation, tracked here: https://github.com/numenta/nupic.core/issues/974

### PR DESCRIPTION
**Pull request by hernandezurbina**
Fixes #3148

### CODE REVIEW

**Comment by rhyolight** [comment]
@hernandezurbina Hey Victor, please have a quick read: https://github.com/numenta/nupic/wiki/Development-Process

You must create a public issue describing the work that needs to be done. Then link to it from this PR description with `fixes #XXX`.

**Comment by subutai** [comment]
@hernandezurbina There is already an existing issue that you should link to: 
https://github.com/numenta/nupic/issues/3148

**Comment by subutai** [comment]
@hernandezurbina Please see the details of the #3148 - you need to handle backward compatibility with pickling. @mrcslws might be able to help you with the details.

**Comment by hernandezurbina** [comment]
@rhyolight @subutai @mrcslws I think it's done. Could you please have a look? Cheers!

File: src/nupic/research/spatial_pooler.py [hunk 1]
```diff
@@ -717,6 +719,14 @@ def getConnectedCounts(self, connectedCounts):
     'connectedCounts' size must match the number of columns"""
     connectedCounts[:] = self._connectedCounts[:]
 
+  def getOverlaps(self):
+    """Returns the overlap score for each column."""
+    return self._overlaps
+
```
Comment (subutai) [inline src/nupic/research/spatial_pooler.py]: Two blank lines between methods.

File: src/nupic/research/spatial_pooler.py [hunk 2]
```diff
@@ -1716,7 +1731,7 @@ def write(self, proto):
 
     boostFactorsProto = proto.init("boostFactors", len(self._boostFactors))
     for i, v in enumerate(self._boostFactors):
-      boostFactorsProto[i] = float(v)
+      boostFactorsProto[i] = float(v)
```
Comment (subutai) [inline src/nupic/research/spatial_pooler.py]: Remove trailing whitespace

File: tests/unit/nupic/research/spatial_pooler_unit_test.py [hunk 3]
```diff
@@ -151,6 +151,39 @@ def testCompute2(self):
       perm = sp._permanences.getRow(columnIndex)
       self.assertEqual(list(perm), list(potential))
```
Comment (subutai) [inline tests/unit/nupic/research/spatial_pooler_unit_test.py]: Two blank lines.

File: tests/unit/nupic/research/spatial_pooler_unit_test.py [hunk 4]
```diff
@@ -151,6 +151,39 @@ def testCompute2(self):
...
+    	numActiveColumnsPerInhArea=5,
+    	globalInhibition=True,
+    	seed=1,
+    	synPermActiveInc=0.1,
+    	synPermInactiveDec=0.1)
+    
+    inputVector = numpy.ones(5)
+    activeArray = numpy.zeros(3)
+    
+    expOutput = numpy.array([2, 0, 0], dtype="float32")
```
Comment (subutai) [inline tests/unit/nupic/research/spatial_pooler_unit_test.py]: dtype should be realDType from spatial_pooler.py

File: tests/unit/nupic/research/spatial_pooler_unit_test.py [hunk 5]
```diff
@@ -151,6 +151,39 @@ def testCompute2(self):
...
+    
+    boostFactors = 2.0 * numpy.ones(3)
+    
+    sp.setBoostFactors(boostFactors)
+
```
Comment (subutai) [inline tests/unit/nupic/research/spatial_pooler_unit_test.py]: Don't need so many blank lines between individual statements.

**Comment by subutai** [comment]
@hernandezurbina Looks good. Left a few minor comments. Please read through our Python style guide: 
https://github.com/numenta/nupic/wiki/Python-Style-Guide

File: src/nupic/research/spatial_pooler.py [hunk 6]
```diff
@@ -1645,6 +1655,11 @@ def __setstate__(self, state):
       # the wrapAround property was added in version 2,
       # in version 1 the wrapAround parameter was True for SP initialization
       state['_wrapAround'] = True
+    if state['_version'] < 3:
+      # the overlaps and boostedOverlaps properties were added in version 3,
+      state['_overlaps'] = numpy.zeros(self._numColumns)
+      state['_boostedOverlaps'] = numpy.zeros(self._numColumns)
```
Comment (mrcslws) [inline src/nupic/research/spatial_pooler.py]: We should probably also set the `dtype`s here to be consistent.

**Comment by hernandezurbina** [comment]
@subutai OK, it's fixed now. Now I'm reading the Python style guide.

**Comment by mrcslws** [comment]
Really, we don't need to initialize these numpy arrays. We could just set `self._overlaps` and `self._boostedOverlaps` to `None`. We build a new array every timestep, so these arrays are never actually used. If we allocate an array in the `__init__` method, it implies that we're going to reuse it.

This would save us from having `numpy.zeros` calls all over the place. Whether we reuse the arrays or not, there only needs to be one code location that calls `numpy.zeros` (or rather, one for each of the two arrays).

**Comment by hernandezurbina** [comment]
@subutai @mrcslws OK, I added the dtype specifications, and left the numpy.zeros calls in __init__ as discussed.

**Comment by subutai** [comment]
@rhyolight Did this PR pass all checks? It only shows the appveyor check.  Thanks.

@mrcslws Has Victor addressed all your comments? I'm fine with the current status.

**Comment by mrcslws** [comment]
👍

**Comment by rhyolight** [comment]
@subutai That's strange... No it didn't. I'm not sure why. When something like this happens, I usually have to push a dummy commit to the remote branch to get the builds to re-trigger. @hernandezurbina You can create a dummy commit like this:

    git commit -m "Trigger CI" --allow-empty
    git push origin RES217

**Comment by hernandezurbina** [comment]
@rhyolight done

**Comment by rhyolight** [comment]
Still strange... looks like the CI servers are running, but the local validations are not. This smacks of webhook latency from GitHub.

**Comment by rhyolight** [comment]
There we go. I forced a validation manually [from the tooling server](http://tooling.numenta.org/validate?repo=numenta/nupic&sha=92aa1d2094f74419f53f445af9244142fb2d6b87).

**Comment by hernandezurbina** [comment]
:thumbsup:

**Pull request merged**